File: backend/mr-db-sql/src/main/java/io/featurehub/db/publish/CacheSource.java
package io.featurehub.db.publish;

import io.featurehub.db.model.DbApplicationFeature;
import io.featurehub.db.model.DbEnvironment;
import io.featurehub.db.model.DbFeatureValue;
import io.featurehub.db.model.DbRolloutStrategy;
import io.featurehub.db.model.DbServiceAccount;
import io.featurehub.dacha.model.PublishAction;
import org.jetbrains.annotations.NotNull;

import java.util.UUID;

public interface CacheSource {
  void registerCache(String cacheName, CacheBroadcast cacheBroadcast);
  void publishToCache(String cacheName);

  void deleteFeatureChange(DbApplicationFeature feature, UUID environmentId);

  /**
   * Whenever a service account changes or at start up, publish it
   */
  void updateServiceAccount(DbServiceAccount serviceAccount, PublishAction publishAction);

  void deleteServiceAccount(UUID id);

  /**
   * whenever an environment changes or at start up, publish it.
   *
   * This is used to cache environments and their feature values.
   */
  void updateEnvironment(DbEnvironment environment, PublishAction publishAction);

  void deleteEnvironment(UUID id);

  void publishFeatureChange(@NotNull DbApplicationFeature appFeature, @NotNull PublishAction update);

  /*
   * This one is used by the Archive to preserve the original feature key
   */
  void publishFeatureChange(@NotNull DbApplicationFeature appFeature, @NotNull PublishAction update,
                            @NotNull String featureKey);
  /**
   * Whenever a feature value changes in any way, publish it out.
   */
  void publishFeatureChange(@NotNull DbFeatureValue strategy);


  void publishRolloutStrategyChange(DbRolloutStrategy rs);
}

File: backend/mr-eventing-common/src/main/kotlin/io/featurehub/mr/events/common/CacheSource.kt
package io.featurehub.mr.events.common

import io.featurehub.dacha.model.PublishAction
import io.featurehub.db.model.*
import java.util.*

interface CacheSource {

  /**
   * publish all environments, service accounts and features associated with this cache
   */
  fun publishObjectsAssociatedWithCache()

  fun deleteFeatureChange(feature: DbApplicationFeature, environmentId: UUID)

  /**
   * Whenever a service account changes or at start up, publish it
   */
  fun updateServiceAccount(serviceAccount: DbServiceAccount, publishAction: PublishAction)

  fun deleteServiceAccount(id: UUID)

  /**
   * whenever an environment changes or at start up, publish it.
   *
   * This is used to cache environments and their feature values.
   */
  fun updateEnvironment(environment: DbEnvironment, publishAction: PublishAction)

  fun deleteEnvironment(id: UUID)


  fun publishFeatureChange(appFeature: DbApplicationFeature, action: PublishAction)

  /*
   * This one is used by the Archive to preserve the original feature key
   */
  fun publishFeatureChange(
      appFeature: DbApplicationFeature, update: PublishAction,
      featureKey: String
  )

  /**
   * Whenever a feature value changes in any way, publish it out.
   */
  fun publishFeatureChange(featureValue: DbFeatureValue)


  fun publishApplicationRolloutStrategyChange(rs: DbApplicationRolloutStrategy)
}
